akka-cluster/src/main/scala/akka/cluster/ClusterDaemon.scala

```diff
@@ -545,16 +559,30 @@ private[cluster] class ClusterCoreDaemon(publisher: ActorRef, joinConfigCompatCh
 
       joinConfigCompatChecker.check(joiningNodeConfig, configWithoutSensitiveKeys) match {
         case Valid ⇒
-          val nonSensitiveKeys = JoinConfigCompatChecker.removeSensitiveKeys(joiningNodeConfig, cluster.settings)
-          // Send back to joining node a subset of current configuration
-          // containing the keys initially sent by the joining node minus
-          // any sensitive keys as defined by this node configuration
-          val clusterConfig = JoinConfigCompatChecker.filterWithKeys(nonSensitiveKeys, context.system.settings.config)
-          sender() ! InitJoinAck(selfAddress, CompatibleConfig(clusterConfig))
+          if (joiningNodeConfig.isEmpty) {
+            // Joining node has version 2.5.9 or earlier.
+            sender() ! InitJoinAck(selfAddress, ConfigCheckUnsupportedByJoiningNode)
+          } else {
+            val nonSensitiveKeys = JoinConfigCompatChecker.removeSensitiveKeys(joiningNodeConfig, cluster.settings)
+            // Send back to joining node a subset of current configuration
+            // containing the keys initially sent by the joining node minus
+            // any sensitive keys as defined by this node configuration
+            val clusterConfig = JoinConfigCompatChecker.filterWithKeys(nonSensitiveKeys, context.system.settings.config)
+            sender() ! InitJoinAck(selfAddress, CompatibleConfig(clusterConfig))
+          }
         case Invalid(messages) ⇒
           // messages are only logged on the cluster side
-          log.warning("Found incompatible settings when [{}] tried to join: {}", sender().path.address, messages.mkString(", "))
-          sender() ! InitJoinAck(selfAddress, IncompatibleConfig)
+          log.warning(
+            "Found incompatible settings when [{}] tried to join: {}. " +
+              "Self version [{}], Joining version [{}].",
+            sender().path.address, messages.mkString(", "),
+            context.system.settings.ConfigVersion, joiningNodeVersion)
+          if (joiningNodeConfig.isEmpty) {
+            // Joining node has version 2.5.9 or earlier.
+            sender() ! InitJoinAck(selfAddress, ConfigCheckUnsupportedByJoiningNode)
+          } else {
+            sender() ! InitJoinAck(selfAddress, IncompatibleConfig)
+          }
```

Comment:
Was thinking maybe would be nice to make a private method for the replies since the
```
            // Joining node has version 2.5.9 or earlier.
            sender() ! InitJoinAck(selfAddress, ConfigCheckUnsupportedByJoiningNode)
```

is in two places; we could have a method that creates the right reply depending on version; otherwise nice

Reply:
Yeah, soudnds good